internal/controller/pkg/revision/dependency.go

```diff
@@ -137,6 +137,24 @@ func (m *PackageDependencyManager) Resolve(ctx context.Context, pkg runtime.Obje
 		Dependencies: sources,
 	}
 
+	// Delete packages in lock with same name and distinct source
+	// This is a corner case when source is updated but image SHA is not (i.e. relocate same image
+	// to another registry)
+	for _, lp := range lock.Packages {
+		if self.Name == lp.Name &&
+			self.Type == lp.Type &&
+			self.Source != lp.Identifier() {
+			if err := m.RemoveSelf(ctx, pr); err != nil {
+				return found, installed, invalid, err
+			}
+			// refresh the lock to be in sync with the contents
+			if err = m.client.Get(ctx, types.NamespacedName{Name: lockName}, lock); err != nil {
+				return found, installed, invalid, err
+			}
```

Comment:
I wonder whether it'd make sense for `RemoveSelf` to take a pointer to the lock as an argument. That way we wouldn't need to fetch it here afterward.

Reply:
RemoveSelf is doing lock get + update, so it is true that this is adding an extra get that could be avoided by returning the lock from RemoveSelf (I believe with current implementation it would make more sense to return the lock rather that taking it as argument).

At the same time, this block of code won't be generally executed, as it only handles corner-case of relocate same image to different registry. Because of this, I decided to keep the changes at minimum, and not change the method signature, but I'd agree with either keeping it as it is or returning the lock in the method, whichever option you prefer.

Reply:
Let's leave it as is. The extra get here is a bit unfortunate but I agree it's not worth the churn of updating the signature at this point.